File: python/cote/GJM/요새.py
from typing import *

x = []
y = []
r = []
longest = 0

class Node :
	def __init__(self) :
		self.children = []
		pass

def square_distance(i: int, j : int) -> int :
	return pow(x[i]-x[j],2) + pow(y[i]- y[j],2)

def contains( i: int, j :int, n: int ) -> bool :
	return r[i] > r[j] and square_distance(i,j) < pow(r[i]-r[j],2)

def get_tree(i : int, n : int) -> Node :
	node = Node()

	for j in range(n) :
		if not contains(i, j, n) :
			continue;

		isChild = True

		for k in range(n) :
			if contains(i,k,n) and contains(k,j,n) :
				isChild = False

		if (isChild) :
			node.children.append(get_tree(j, n))

	return node

def height( cur : Node) -> int :
	global longest 
	if len(cur.children) == 0 :
		return 0


	heights = [] 

	for child in cur.children :
		heights.append(height(child))

	heights.sort()

	if len(heights) >= 2 :
		longest = max(longest , heights[-1] + heights[-2] +2)


	return heights[-1] + 1;

def solve(n : int) -> int :
	global longest
	root = get_tree(0, n)

	h = height(root)

	return max(h, longest);

if __name__ == "__main__" :

	with open("data/요새.txt") as fout :
		TEST_CASES = int(fout.readline())

		for _ in range(TEST_CASES) :
			longest = 0
			n = int(fout.readline())
			forts = [ list(map(int,fout.readline().split(" "))) for _ in range(n)] 
			x = [fort[0] for fort in forts]
			y = [fort[1] for fort in forts]
			r = [fort[2] for fort in forts]
			print(solve(n))

File: c++/cote/GJM/요새.cpp
#include <cmath>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

struct Node {
  vector<Node*> children;
  int i;

  Node(int i) : i(i) {}

  ~Node() {
    for (Node* child : children) {
      delete child;
    }
  }
};

int longest = 0;

int* x;
int* y;
int* radius;

int* getFort(string& s) {
  int* ret = new int[3];
  stringstream ss(s);
  string buf;

  for (int i = 0; i < 3; i++) {
    getline(ss, buf, ' ');
    ret[i] = stoi(buf);
  }

  return ret;
}

int square_distance(int parent, int child) {
  return pow(x[parent] - x[child], 2) + pow(y[parent] - y[child], 2);
}

bool encolses(int parent, int child) {
  return radius[parent] > radius[child] &&
         square_distance(parent, child) <
             pow(radius[parent] - radius[child], 2);
}

bool isChild(int parent, int child, int n) {
  if (!encolses(parent, child)) return false;
  for (int i = 0; i < n; i++) {
    if (i == parent || i == child) continue;
    if (encolses(parent, i) && encolses(i, child)) return false;
  }

  return true;
}

Node* getTree(int i, int n) {
  Node* ret = new Node(i);

  for (int j = 0; j < n; j++) {
    if (isChild(i, j, n)) {
      ret->children.push_back(getTree(j, n));
    }
  }

  return ret;
}

int height(Node* root) {
  vector<int> heights;

  for (int i = 0; i < root->children.size(); i++) {
    heights.push_back(height(root->children[i]));
  }

  if (heights.empty()) return 0;

  sort(begin(heights), end(heights));

  int l = heights.size();
  if (l >= 2) longest = max(longest, 2 + heights[l - 2] + heights[l - 1]);

  return heights.back() + 1;
}

int solve(int n) {
  longest = 0;
  Node* root = getTree(0, n);
  int h = height(root);

  delete root;
  return max(h, longest);
}

int main(void) {
  fstream fd("data/요새.txt");

  string buf;

  getline(fd, buf);
  int test_cases = stoi(buf);

  while (test_cases > 0) {
    getline(fd, buf);
    int n = stoi(buf);

    x = new int[n];
    y = new int[n];
    radius = new int[n];

    for (int i = 0; i < n; i++) {
      getline(fd, buf);
      int* fort = getFort(buf);
      x[i] = fort[0];
      y[i] = fort[1];
      radius[i] = fort[2];
      delete[] fort;
    }

    cout << solve(n) << endl;

    delete[] x;
    delete[] y;
    delete[] radius;

    test_cases--;
  }

  return 0;
}